python/src/main/java/org/apache/zeppelin/python/IPythonInterpreter.java

```diff
@@ -0,0 +1,276 @@
+    } catch (Exception e) {
+      LOGGER.error("Fail to open IPythonInterpreter", e);
+    }
+  }
+
+  private void setupJVMGateway(int jvmGatewayPort) throws IOException {
+    gatewayServer = new GatewayServer(this,
+        jvmGatewayPort,
+        GatewayServer.DEFAULT_PYTHON_PORT,
+        InetAddress.getByName("0.0.0.0"),
+        InetAddress.getByName("0.0.0.0"),
+        GatewayServer.DEFAULT_CONNECT_TIMEOUT,
+        GatewayServer.DEFAULT_READ_TIMEOUT,
+        (List) null);
+    gatewayServer.start();
+
+    InputStream input =
+        getClass().getClassLoader().getResourceAsStream("grpc/python/zeppelin_python.py");
+    List<String> lines = IOUtils.readLines(input);
+    ExecuteResponse response = ipythonClient.execute(ExecuteRequest.newBuilder()
+        .setCode(StringUtils.join(lines, System.getProperty("line.separator"))
+            .replace("${JVM_GATEWAY_PORT}", jvmGatewayPort + "")).build());
+    if (response.getStatus() == ExecuteStatus.ERROR) {
+      throw new IOException("Fail to setup JVMGateway\n" + response.getTextOutput());
+    }
+  }
+
+
+  private void launchIPythonKernel(int ipythonPort)
+      throws IOException, URISyntaxException {
+    File tmpPythonScriptFolder = Files.createTempDirectory("zeppelin_ipython").toFile();
+    String[] ipythonScripts = {"ipython_server.py", "ipython_pb2.py", "ipython_pb2_grpc.py"};
+    for (String ipythonScript : ipythonScripts) {
+      URL url = getClass().getClassLoader().getResource("grpc/python"
+          + "/" + ipythonScript);
+      FileUtils.copyURLToFile(url, new File(tmpPythonScriptFolder, ipythonScript));
+    }
+
+    CommandLine cmd = CommandLine.parse(pythonExecutable);
+    cmd.addArgument(tmpPythonScriptFolder.getAbsolutePath() + "/ipython_server.py");
+    cmd.addArgument(ipythonPort + "");
+    DefaultExecutor executor = new DefaultExecutor();
+    ProcessLogOutputStream processOutput = new ProcessLogOutputStream(LOGGER);
+    executor.setStreamHandler(new PumpStreamHandler(processOutput));
+    watchDog = new ExecuteWatchdog(ExecuteWatchdog.INFINITE_TIMEOUT);
+    executor.setWatchdog(watchDog);
+    executor.execute(cmd, EnvironmentUtils.getProcEnvironment(), this);
+
+    // wait until IPython kernel is started or timeout
+    long startTime = System.currentTimeMillis();
+    while (true) {
+      try {
+        Thread.sleep(100);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
+
+      try {
+        StatusResponse response = ipythonClient.status(StatusRequest.newBuilder().build());
+        if (response.getStatus() == IPythonStatus.RUNNING) {
+          LOGGER.info("IPython Kernel is Running");
+          break;
+        } else {
+          LOGGER.info("Wait for IPython Kernel to be started");
+        }
+      } catch (Exception e) {
+        // ignore the exception, because is may happen when grpc server has not started yet.
+        LOGGER.info("Wait for IPython Kernel to be started");
```

Comment:
As the comment mentions, we can ignore this exception, because it may happen when grpc server has not started yet.